go/vt/vtctl/reparentutil/durability_funcs.go

```diff
@@ -0,0 +1,90 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reparentutil
+
+import (
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	"vitess.io/vitess/go/vt/topo/topoproto"
+	"vitess.io/vitess/go/vt/vtctl/reparentutil/promotionrule"
+)
+
+// SemiSyncAckersForPrimary returns the list of tablets which are capable of sending Semi-Sync Acks for the given primary tablet
+func SemiSyncAckersForPrimary(primary *topodatapb.Tablet, allTablets []*topodatapb.Tablet) (semiSyncAckers []*topodatapb.Tablet) {
+	for _, tablet := range allTablets {
+		if topoproto.TabletAliasEqual(primary.Alias, tablet.Alias) {
+			continue
+		}
+		if IsReplicaSemiSync(primary, tablet) {
+			semiSyncAckers = append(semiSyncAckers, tablet)
+		}
+	}
+	return
+}
+
+// RevokeForTablet checks whether we have reached enough tablets such that the given primary eligible tablet cannot accept any new writes
+func RevokeForTablet(primaryEligible *topodatapb.Tablet, tabletsReached []*topodatapb.Tablet, allTablets []*topodatapb.Tablet) bool {
+	// if we have reached the primaryEligible tablet and stopped its replication, then it will not
+	// accept any new writes
+	if topoproto.TabletInList(primaryEligible, tabletsReached) {
+		return true
+	}
+
+	// semiSyncAckersReached is the list of reachable tablets capable of sending semi sync Acks for the given primaryEligible tablet
+	semiSyncAckersReached := SemiSyncAckersForPrimary(primaryEligible, tabletsReached)
+
+	// allSemiSyncAckers is the list of reachable tablets capable of sending semi sync Acks for the given primaryEligible tablet
+	allSemiSyncAckers := SemiSyncAckersForPrimary(primaryEligible, allTablets)
+
+	// numOfSemiSyncAcksRequired is the number of semi sync Acks that the primaryEligible tablet requires
+	numOfSemiSyncAcksRequired := SemiSyncAckers(primaryEligible)
+
+	// if we have reached enough semi-sync Acking tablets such that the primaryEligible cannot accept a write
+	// we have revoked from the tablet
+	return len(allSemiSyncAckers)-len(semiSyncAckersReached) < numOfSemiSyncAcksRequired
+}
+
+// Revoked checks whether we have reached enough tablets to guarantee that no tablet eligible to become a primary can accept any write
+func Revoked(tabletsReached []*topodatapb.Tablet, allTablets []*topodatapb.Tablet) bool {
+	for _, tablet := range allTablets {
+		if PromotionRule(tablet) == promotionrule.MustNot {
+			continue
+		}
+		if !RevokeForTablet(tablet, tabletsReached, allTablets) {
+			return false
+		}
+	}
+	return true
+}
+
+// EstablishForTablet checks whether we have reached enough tablets to say that the given primary eligible tablet will be able to accept new writes
+func EstablishForTablet(primaryEligible *topodatapb.Tablet, tabletsReached []*topodatapb.Tablet) bool {
```

Comment:
Some of the feedback from the base PR is also valid here..
Whether this should be exported (may be not), prefix the name with `Is` etc.